### ISSUE TITLE
[cryptolib] HW backed keys seems be misusing keymgr

### ISSUE THREAD

**Issue by vsukhoml** [lowRISC/opentitan#22624]
### Description

KEYMGR states [salt reg](https://opentitan.org/book/hw/ip/keymgr/doc/registers.html#salt) is 256-bit, and there is [version reg](https://opentitan.org/book/hw/ip/keymgr/doc/registers.html#key_version), but [hw backed key implementation](https://github.com/lowRISC/opentitan/blob/master/sw/device/lib/crypto/include/key_transport.h#L74) use 224 bit input, and puts version [into word 0](https://github.com/lowRISC/opentitan/blob/master/sw/device/lib/crypto/impl/key_transport.c#L77)

Either documentation or code is incorrect and should be fixed.

@moidx

**Comment by moidx** [comment]
The versioning scheme requires using the version register. This is required for use cases relying on the keymgr anti rollback scheme. 

The salt supported by hardware is 256bit, and it would be preferable to support this at the API level as this is something that aligns better with existing applications.

**Comment by jadephilipoom** [comment]
The keyblob doesn't get directly passed to keymgr; the version put into the version register. However, we still only make 7 words of diversification available via the API, because [the last word of the diversification is set to the key mode](https://github.com/lowRISC/opentitan/blob/641d7b11a82a9cc69360642d139db9c236f6c072/sw/device/lib/crypto/impl/keyblob.c#L108). This ensures you can't use the same key material for keys of different modes (for example, using the same seed for ECDSA-P256 and ECDSA-P384, which would weaken security guarantees for the P-384 usage).

**Comment by jadephilipoom** [comment]
If we want to optionally support a 256-bit salt and just say in usage guidance that the salt should always be different for different modes, I'm OK with that. We could for example allow keyblobs for hardware-backed keys to be either 8 or 9 words long (version + 8 words of salt) and if we see a 9-word keyblob on a hardware-backed key then we could skip the mode in the salt.

A potentially more elegant option is to use a hash as the salt; this is standard for HMAC-KDF if the salt is longer than the HMAC key size, but is not usually necessary for KMAC-KDF because it can handle arbitrary-length keys. Then we could allow the user to specify any length of diversification data, tack the mode onto the end, and hash the whole thing if it's too long.

**Comment by ballifatih** [comment]
Don't want to piggy back other issues onto this discussion, but they seem quite related:

- Add cryptolib function to read MAX_KEY_VERSION (#22214). This makes me wonder, who actually sets `KEY_VERSION` and to which value during key generation. This value should be related to the version of the SW. Is the user aware that they need to keep track of the SW version?
- Reorganize diversification inputs (#22565). What @jadephilipoom needs is a way to separate different cryptographic use cases. This is actually what the second 256-bit is supposed to do in the message payload in RTL and this is properly done for hw-backed keys (AES/KMAC/OTBN), but currently this is not exposed for SW (which uses `NONE` as safe default). Current HW implementation only has 4 values, so it cannot be used to solve the SW needs. In the future, it would be nice to expose this as a CSR to SW so that 256-bit `key_mode` besides 256-bit `salt` left to the user.

As a note to @vsukhom, the function to obtain SW key from keymgr is not yet implemented, so this is not the function you need.

**Comment by moidx** [comment]
Hi @jadephilipoom, thanks for the context. My preference would be to support a hash as the salt. We are moving in this direction in cases where we need to aggregate measurements into a bit width supported by `keymgr` inputs.

I also agree that it would be good for the cryptolib to combine the key type identifier to deter misuse.

**Comment by ballifatih** [comment]
> Are there any real uses for `version` in keymgr?

For Keymgr HW, `KEY_VERSION` register has a specific meaning and function. Its unsigned integer value is compared against one of `MAX_<STAGE>_KEY_VER_SHADOWED` registers (depending on the Keymgr `<STAGE>`), and the key generation request to HW only succeeds if `KEY_VERSION <= MAX_<STAGE>_KEY_VER_SHADOWED`. So, we should not use this register for `key_mode`.

**Comment by jadephilipoom** [comment]
I think there's some confusion -- we are definitely not using the version register for `key_mode`.

The version register gets the version (it's stored in the first word of the keyblob when in blinded key form, but this is an implementation detail that callers shouldn't have to reason about). The key mode currently goes as the last word of the salt input to keymgr.

**Comment by vsukhoml** [comment]
So, before chips go in production we need to stick a method for hw backed key generation such that:
1. hw backed keys will remain the same on the same hardware, given same salt. This property seems to be maintained.
2. the method used for HW backed keys should be FIPS approved - see [#22297](https://github.com/lowRISC/opentitan/issues/22297#issuecomment-2066946326)
3. provide diversification for keys created for different algorithms & modes (can be split responsibility of sw and hw, as anyway firmware will be part of certification boundary)

Specific implementation of №3 should be defined early, if hw backed keys used for attestation, so they can be recreated later. But cryptolib may offload this responsibility to firmware which can add many more properties like `key purpose`, `application id`, etc not available at lower level.  This is where a desire to have a generic API with plain 256-bits comes from --- enable specific uses to construct key salts as they wish. In existing API it is just an inconvenience - 7 words in memory, 1 word as `version` argument. I see no reason why not to make it all 8 words in memory - simplify code, and caller can use word 0 as version if desired.

**Comment by vsukhoml** [comment]
@jadephilipoom , yes, there was definitely misunderstanding from my side - I saw code [key->keyblob[0] = version;](https://github.com/lowRISC/opentitan/blob/master/sw/device/lib/crypto/impl/key_transport.c#L77)  but confused with what happens after it. Now, tracing further I found [diversification->version = launder32(keyblob[0]);](https://github.com/lowRISC/opentitan/blob/master/sw/device/lib/crypto/impl/keyblob.c#L102), and setting the key mode as last word of the constant loaded as `salt`.  

Yes, caller have to be "responsible" in what it does. I'm fine with hw-backed keys being implemented this way, assuming there is no other hw mechanism to diversify keys for different users (AES, KMAC, etc).

> For that reason I prefer the solution of prehashing the mode followed by the diversification data; then the caller has even more flexibility and can use as much diversification data as they want.

Yes, this is most flexible, though it adds performance overhead each time key is constructed. I think it is ok to stay with existing implementation now.  It is the documentation & code comments that shall be updated to avoid confusion.

**Comment by johannheyszl** [comment]
seems obsolote
